bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/WatchdogInterceptor.java

```diff
@@ -0,0 +1,208 @@
+      WatchedCall<?,?> stream = it.next().getKey();
+      if (stream.cancelIfStale()) {
+        it.remove();
+      }
+    }
+  }
+
+  enum State {
+    /** Stream has been started, but doesn't have any outstanding requests. */
+    IDLE,
+    /** Stream is awaiting a response from upstream. */
+    WAITING,
+    /**
+     * Stream received a response from upstream, and is awaiting outerResponseObserver processing.
+     */
+    DELIVERING
+  }
+
+  class WatchedCall<ReqT, RespT> extends SimpleForwardingClientCall<ReqT, RespT> {
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
+    private State state;
+
+    @GuardedBy("lock")
+    private int pendingCount = 0;
+
+    @GuardedBy("lock")
+    private long lastActivityAt = clock.millisTime();
+
+    WatchedCall(ClientCall<ReqT, RespT> delegate) {
+      super(delegate);
+    }
+
+    @Override
+    public void start(Listener<RespT> listener, Metadata metadata) {
+      ClientCall<ReqT, RespT> call = delegate();
+
+      synchronized (lock) {
+        Preconditions.checkState(state == null, "Already started");
+        state = (pendingCount == 0) ? State.IDLE : State.WAITING;
+        lastActivityAt = clock.millisTime();
+      }
+
+      openStreams.put(this, PRESENT);
+
+      call.start(new SimpleForwardingClientCallListener<RespT>(listener) {
+        @Override
+        public void onMessage(RespT message) {
+          synchronized (lock) {
+            state = State.DELIVERING;
+          }
+
+          super.onMessage(message);
+
+          synchronized (lock) {
+            pendingCount--;
+            lastActivityAt = clock.millisTime();
+
+            state = (pendingCount > 0) ? State.WAITING : State.IDLE;
+          }
+        }
+
+        @Override
+        public void onClose(Status status, Metadata trailers) {
+          openStreams.remove(WatchedCall.this);
+
+          // Take care to convert an error caused by StreamWaitTimeoutException into an ABORTED
+          // status. This will allow it to be retried.
+          if (status.getCause() instanceof StreamWaitTimeoutException) {
+            status = Status.ABORTED
+                .withDescription(status.getDescription())
+                .withCause(status.getCause());
+          }
+
+          super.onClose(status, trailers);
+        }
+      }, metadata);
+    }
+
+    @Override
+    public void request(int count) {
+      synchronized (lock) {
+        if (state == State.IDLE) {
+          state = State.WAITING;
+          lastActivityAt = clock.millisTime();
+        }
+
+        // Increment the request count without overflow
+        int maxIncrement = Integer.MAX_VALUE - pendingCount;
+        count = Math.min(maxIncrement, count);
+        pendingCount += count;
+      }
+
+      super.request(count);
+    }
+
+    boolean cancelIfStale() {
+      synchronized (lock) {
+        long waitTime = clock.millisTime() - lastActivityAt;
+
+        switch (this.state) {
+          case IDLE:
```

Comment:
please add a case for the other states, which can serve as documentation.  A default handling is a good practice as well.

Reply:
will do